src/Features/Core/Portable/RemoveUnusedParametersAndValues/AbstractRemoveUnusedParametersAndValuesCodeFixProvider.cs

```diff
@@ -0,0 +1,702 @@
+                Debug.Assert(removeAssignments == AbstractRemoveUnusedParametersAndValuesDiagnosticAnalyzer.GetIsRemovableAssignmentDiagnostic(diagnostic));
+            }
+#endif
+
+            return GetDiagnosticsGroupedByMember(diagnostics, syntaxFacts, root);
+        }
+
+        private IEnumerable<IGrouping<SyntaxNode, Diagnostic>> GetDiagnosticsGroupedByMember(
+            ImmutableArray<Diagnostic> diagnostics,
+            ISyntaxFactsService syntaxFacts,
+            SyntaxNode root)
+            => diagnostics.GroupBy(d => syntaxFacts.GetContainingMemberDeclaration(root, d.Location.SourceSpan.Start));
+
+        protected override async Task<Document> FixAllAsync(Document document, ImmutableArray<Diagnostic> diagnostics, CancellationToken cancellationToken)
+        {
+            // Track all the member declaration nodes that have diagnostics.
+            // We will post process all these tracked nodes after applying the fix (see "PostProcessDocumentAsync" below in this source file).
+
+            var syntaxFacts = document.GetLanguageService<ISyntaxFactsService>();
+            var root = await document.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+            var memberDeclarations = GetDiagnosticsGroupedByMember(diagnostics, syntaxFacts, root).Select(g => g.Key);
+            root = root.ReplaceNodes(memberDeclarations, computeReplacementNode: (_, n) => n.WithAdditionalAnnotations(s_memberAnnotation));
+            document = document.WithSyntaxRoot(root);
+            return await base.FixAllAsync(document, diagnostics, cancellationToken).ConfigureAwait(false);
+        }
+
+        protected sealed override async Task FixAllAsync(Document document, ImmutableArray<Diagnostic> diagnostics, SyntaxEditor editor, CancellationToken cancellationToken)
+        {
+            var root = await document.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+            var syntaxFacts = document.GetLanguageService<ISyntaxFactsService>();
+            var diagnosticsGroupedByMember = GetDiagnosticsGroupedByMember(diagnostics, syntaxFacts, root,
+                out var diagnosticId, out var preference, out var removeAssignments);
+            if (preference == UnusedValuePreference.None)
+            {
+                return;
+            }
+
+            var semanticModel = await document.GetSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+            var usedNames = PooledHashSet<string>.GetInstance();
+            try
+            {
+                foreach (var diagnosticsToFix in diagnosticsGroupedByMember)
+                {
+                    var orderedDiagnostics = diagnosticsToFix.OrderBy(d => d.Location.SourceSpan.Start);
+                    FixAll(diagnosticId, orderedDiagnostics, semanticModel, root, preference,
+                        removeAssignments, GenerateUniqueNameAtSpanStart, editor, syntaxFacts, cancellationToken);
```

Comment:
the usage of passing along local functoins to other methods that then mutate code within this block, combined with zero comments explaining any of this flow makes this all exceptionally difficult to understand.  I'm really not a fan of this approach of treating a method body as a class, and needing to understand super complex control and data flow for these types of features.  IMO it just leads to obfuscation and most time spent just trying to piece how things are working.  By the time that is done, it's hard to even tell if it's doing the right thing.

Reply:
Refactored into a separate type `UniqueVariableNameGenerator`.

---
In reply to: [230604387](https://github.com/dotnet/roslyn/pull/30777#discussion_r230604387) [](ancestors = 230604387)